allennlp/training/util.py

```diff
@@ -428,28 +418,38 @@ def make_vocab_from_params(
             "The 'vocabulary' directory in the provided serialization directory is non-empty"
         )
 
-    all_datasets = datasets_from_params(params)
-    datasets_for_vocab_creation = set(params.pop("datasets_for_vocab_creation", all_datasets))
-
-    for dataset in datasets_for_vocab_creation:
-        if dataset not in all_datasets:
-            raise ConfigurationError(f"invalid 'dataset_for_vocab_creation' {dataset}")
-
-    logger.info(
-        "From dataset instances, %s will be considered for vocabulary creation.",
-        ", ".join(datasets_for_vocab_creation),
+    datasets_for_vocab_creation: Optional[List[str]] = params.pop(
+        "datasets_for_vocab_creation", None
     )
+    # Do a quick sanity check here. There's no need to load any datasets if the vocab
+    # type is "empty".
+    if datasets_for_vocab_creation is None and vocab_params.get("type") == "empty":
```

Comment:
I'm not crazy about this ad hoc solution here but I don't see another way. At least I added a test for this case to make sure we don't get a regression in the future.

Reply:
A possible alternative is to build a lazy generator here that doesn't actually call `.read()` on anything unless it's needed.  Not sure how feasible that is, though.

Reply:
One obvious concern is that it requires changes to `datasets_from_params` also in ways that change the return type of that method; I'm pretty sure, though that this is the only use of `datasets_from_params`, and this function is only ever called on the master process in a distributed setting.  In every other configuration, we go through a different code path.

Reply:
It's also called from the `find_learning_rate` command, but that's it.

Reply:
This is another reason why I think readers should always be lazy. Leave the lazy/non-lazy decision up to the dataset.

Reply:
Yeah, I can see that.  Where do you make the lazy / non-lazy decision, then?  You have to make a choice once you get to the `DataLoader`, because of how batching / sampling works.  You just base the decision of which dataset to use on how you've decided to do data loading?

Reply:
I was either thinking we'd make that decision part of the `DataLoader` API, or have a separate configuration for the type of `Dataset` to use.

I like the first option because the lazy vs. non-lazy is literally an issue of data loading. The downside though is that we'd be breaking from the PyTorch API. But having to separately configure a `DatasetReader`, `Dataset`, and `DataLoader` seems overboard.